### ISSUE TITLE
Eclipse 4: Continue migration [TIMEBOX: 8d]

### ISSUE THREAD

**Issue by AdrianPotter** [ISISComputingGroup/IBEX#2491]
For August 2017 sprint. Continue E4 migration. Details to be filled in by Adrian

**Comment by AdrianPotter** [comment]
I've written some notes on the first stage of migration here:

https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Eclipse-4-migration-notes

For the 2nd stage of the timeboxed migration we should aim to do the following:

- Migrate the alarms view
- Migrate the edit configuration dialog
- Create a proper perspective switcher including reset perspective option

## The alarms view
In order to migrate the alarms view we'll need to sort out the following issues:

- JDBC connection currently isn't working. If we get that sorted, we should be able to fully activate the beam status view.
- Sort out the ActiveMQ connection
- Find a solution to the E3/E4 compatibility issue with CSS views as described in the notes

Please make sure to document how each of these issues were solved. The last one in particular is going to be recurrent.

## Edit configuration dialog
This hopefully should be relatively straightforward since the back end only relies on talking to PVs which we know we can do.

The current mock dialog is based on the E4 application model. That migration may involve translating the current dialog tree into the new application mode rather than relying on the existing code. This will be exploratory in nature

## Create a proper perspective switcher
As described in the notes, the current perspective switcher uses hard-coded perspectives. It'd be good to get a flexible system in place at an early stage that makes it easy to add new perspectives. There may be multiple ways to do this, either via the E4 application model or by using extension points.

The perspective switcher should also allow users to reset a perspective to its original settings. From reading online, this solution may be available to us since we're using pure E4:

https://www.eclipse.org/forums/index.php/t/457640/

I think this is the solution for an application using compatibility at the front end:

https://stackoverflow.com/questions/19717154/how-do-i-reset-perspective-for-eclipse-e4-rcp-application

**Comment by AdrianPotter** [comment]
Notes about the view issue from TechTalk:

>Hi:
>
>> I’m trying to write an Eclipse 4 application using EPICS and CSS.
> 
>> From inspection of the code, CSS’s views use the Eclipse 3.x compatibility layer; they inherit from ViewPart and call creatPartControl() to build the view.
>> Elements of the Eclipse 4 application model on the other hand use > annotations (i.e.  @PostConstruct) to do the same.
>
>> If I try and use a CSS ViewPart in my Eclipse 4 application, I typically get nullPointerExceptions where the CSS class tries to access the Eclipse  3.x API (e.g. getSite()).
>>  Does anyone have any advice on how I can get around this issue  to use CSS views in my Eclipse 4 application?
>
>You can have POD-type contributions with @PostConstruct etc.
>Just search for @PostConstruct in the CSS sources and you'll find the Clock plugin as an example.
>(https://github.com/ControlSystemStudio/cs-studio/blob/master/applications/apputil/apputil-plugins/org.csstudio.utility.clock/src/org/csstudio/utility/clock/ClockView.java ) The associated plugin.xml registers that POD as an "e4view" (https://github.com/ControlSystemStudio/cs-studio/blob/master/applications/apputil/apputil-plugins/org.csstudio.utility.clock/plugin.xml#L10).
>
>For what it's worth, the "compatibility layer" is somewhat misnamed.
>RCP is not going away, E4 is not replacing it.
>A better way to describe it:
>E4 is a low-level model-based description of parts, menues etc. which are then rendered by SWT.
>That's great for building a new standalone browser for your DVD collection, or for a clock.
>For an application you typically also want online help, preference UI, file navigator, ..
>That's what RCP adds on top of E4.
>The "compatibility layer" would better be named "rich application layer" or "workspace  and workbench layer".
>The Eclipse IDE for example will continue to use it.
>See also "Tips and tricks for your Eclipse 4 migration." on https://www.eclipsecon.org/na2016/conference/session-slides 
>
>-Kay
>

**Comment by AdrianPotter** [comment]
@John-Holt-Tessella, I'll start working on the alarms view and hopefully in the process figure out how to put CSS views that use the compatibility layer into our E4 application

**Comment by AdrianPotter** [comment]
This is pretty much done for this timebox. Remaining tasks:

- [ ] Add an icon to the reset perspectives button
- [ ] Make the dashboard and blocks parts shared. I made their containers not shared to avoid issues with resizing being shared between perspectives and the reset button not working on them. I think if the parts themselves are shared though it shouldn't be an issue.
- [ ] Make the DAE view visible and make sure it now resizes properly. I fixed it on the E3 branch and merged so hopefully it's now ok here too.
- [ ] Any tidying of the code we think is necessary so as not to add to the technical debt (TARGET_ID for one I think is repeated in two classes)
- [ ] Write up any findings on the Wiki, including anything @John-Holt-Tessella wants to add
- [ ] Add details to the ticket for next sprint about what is likely to be most useful. My suggestion would be the remaining perspectives starting with the synoptic and device screens. Based on experience from this ticket, there's a good chance it will be relatively straightforward. The ease of the scripting view will depend how cooperative PyDev is. If we run it on the compatibility layer though, hopefully it should work as before.

I'm on holiday for a week so if @John-Holt-Tessella, you wanted to pick it up, please do, but it might be tricky without me around to explain. HINT: Almost everything the application needs to know, including what to run on the compatibility layer, is in the e4Xmi.

**Comment by AdrianPotter** [comment]
Just needs any input from John on:

https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Eclipse-4-migration-notes

**Comment by Tom-Willemsen** [comment]
I've had a look at the changes - after some initial setup steps (which I've asked Adrian to put on the wiki) it works.

Things I've looked at:
- Alarms perspective - looks good, functionality is as before (and looks better)
- DAE perspective - all appears to be functional
- Beam status - The graph appears. @AdrianPotter confirmed that the controls are currently disabled by default so it's non-functional at the moment.
- Configuration dialogs - I've been through the basic functionality and it all seems to work as before.
- Web links perspective - works as before as far as I can see
- Reset layout button - does what I expect
- Perspective switcher - allows me to switch perspectives

I've skimmed over the code and it all looks sensible, I haven't had a super-detailed look as most of it is just E4 metadata
